src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,50 +1,194 @@
+    private Node<T> last;
+
+    public MyLinkedList() {
+    }
+
+    private static class Node<T> {
+        private T element;
+        private Node<T> next;
+        private Node<T> prev;
+
+        public Node(Node<T> prev, T element, Node<T> next) {
+            this.next = next;
+            this.element = element;
+            this.prev = prev;
+        }
+    }
+
     @Override
     public boolean add(T value) {
-        return false;
+        addElement(value);
+        return true;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException(EXCEPTION_MESSAGE);
+        }
+        if (index == size) {
+            addElement(value);
+        } else {
+            addElementByIndex(value, index);
+        }
+    }
 
+    private void addElement(T value) {
+        Node<T> newNode = new Node<>(last, value, null);
+        Node<T> lastForThisNode = last;
+        last = newNode;
+        if (lastForThisNode == null) {
+            first = newNode;
+        } else {
+            lastForThisNode.next = newNode;
+        }
+        size++;
+    }
+
+    private void addElementByIndex(T value, int index) {
+        if (index == 0) {
+            Node<T> newNode = new Node<>(null, value, findByIndex(index));
+            findByIndex(index).prev = newNode;
+            first = newNode;
+            size++;
+        } else {
+            Node<T> after = findByIndex(index);
+            Node<T> before = after.prev;
+            Node<T> newNode = new Node<>(before, value, after);
+            after.prev = newNode;
+            before.next = newNode;
+            size++;
+        }
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        if (list.size() == 0) {
+            return false;
+        }
+        for (T value : list) {
+            add(value);
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return findByIndex(index).element;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> setNode = findByIndex(index);
+        T oldValue = setNode.element;
+        setNode.element = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> removeNode = findByIndex(index);
+        return removeHelper(removeNode);
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        Node<T> currentNode = first;
+        Node<T> removeNode = null;
+        for (int i = 0; i < size; i++) {
+            if (Objects.equals(currentNode.element, object)) {
+                removeNode = currentNode;
+                break;
+            }
+            currentNode = currentNode.next;
+        }
+
+        if (removeNode == null) {
+            return false;
+        }
+        removeHelper(removeNode);
```

Comment:
let's move `removeHelper(removeNode);` into the if block and add `return true` after that. Then we can remove `removeNode ` variable

Reply:
done